### PR TITLE
`ParticleHandler`: use tolerance for is_inside_unit_cell() check

### PR DESCRIPTION
**Pull request by mschreter**
This PR proposes to add a tolerance of `1e-12` for checking if a particle is inside a cell via `GeometryInfo::is_inside_unit_cell()`. The motivation is that we observed a lost particle when the location corresponded to a vertex for the following mesh (tested with `MappingQ(3)`):

image

I've added a test where the misbehavior is triggered and now fixed in this PR. I've verified that the tests inside the `tests/particles` folder still pass.

@j-brotz FYI

### CODE REVIEW

File: include/deal.II/particles/particle_handler.h [hunk 1]
```diff
@@ -1018,6 +1018,11 @@ namespace Particles
      */
     unsigned int handle;
 
+    /**
+     * Tolerance to be used for GeometryInfo<dim>::is_inside_cell().
+     */
+    const double tol = 1e-12;
```
Comment (kronbichler) [inline include/deal.II/particles/particle_handler.h:1024]: Can we use a more verbose variable name, e.g. `tolerance_inside_cell` or something related? I think that would make it clearer for the reader of the relevant places in the `cc` file (without hovering over the variable to see the documentation).

**Review by kronbichler — APPROVED** [review]
I agree, such a tolerance is definitely reasonable.

**Review by bangerth — COMMENTED** [review]
With this change, particles can end up (geometrically) in multiple cells. Have you checked that in all places where you make the change, the code always identifies one unique cell after which it stops searching?

**Comment by kronbichler** [comment]
Your test seems to fail on the serial machine because you use a `parallel::distributed::Triangulation`:
```
/jenkins/workspace/dealii-serial_PR-16691/tests/particles/particle_handler_sort_02.cc: In instantiation of ‘void test() [with int dim = 2; int spacedim = 2]’:
/jenkins/workspace/dealii-serial_PR-16691/tests/particles/particle_handler_sort_02.cc:92:14:   required from here
/jenkins/workspace/dealii-serial_PR-16691/tests/particles/particle_handler_sort_02.cc:36:55: error: use of deleted function ‘dealii::parallel::distributed::Triangulation<dim, spacedim>::Triangulation(MPI_Comm, typename dealii::Triangulation<dim, spacedim>::MeshSmoothing, dealii::parallel::distributed::Triangulation<dim, spacedim>::Settings) [with int dim = 2; int spacedim = 2; MPI_Comm = int; typename dealii::Triangulation<dim, spacedim>::MeshSmoothing = dealii::Triangulation<2, 2>::MeshSmoothing]’
   36 |   parallel::distributed::Triangulation<dim, spacedim> tr(MPI_COMM_WORLD);
      |                                                       ^~
In file included from /jenkins/workspace/dealii-serial_PR-16691/tests/particles/particle_handler_sort_02.cc:19:
/jenkins/workspace/dealii-serial_PR-16691/include/deal.II/distributed/tria.h:1032:16: note: declared here
 1032 |       explicit Triangulation(
      |                ^~~~~~~~~~~~~
....
```
Is this necessary or can you switch to the standard `dealii::Triangulation`? Otherwise, restrict the test to `_with_mpi=true`.

**Comment by mschreter** [comment]
> With this change, particles can end up (geometrically) in multiple cells. Have you checked that in all places where you make the change, the code always identifies one unique cell after which it stops searching?

That's a good point. For our test, the problematic search happened inside `sort_particles_into_subdomains_and_cells()`, where it is guaranteed that only one cell is identified. However, I will verify other occurrences as well.

**Review by gassmoeller — APPROVED** [review]
I agree this is an improvement and the tests seems to be fine. Just for future reference I think #13865 addressed the same problem, but I only fixed it in one place instead of all the relevant places. Thanks for the fix!

**Comment by mschreter** [comment]
> I agree this is an improvement and the tests seems to be fine. Just for future reference I think #13865 addressed the same problem, but I only fixed it in one place instead of all the relevant places. Thanks for the fix!

@gassmoeller great. So far, I did not have a look into the potential corner case that with this fix a particle could be located in multiple cells (as mentioned by @bangerth in https://github.com/dealii/dealii/pull/16691#issuecomment-1961195652). Considering your experience regarding the particle framework, what is your opinion on that? Do you think it is sufficient that the existing tests passed or should we investigate it further?

**Comment by gassmoeller** [comment]
I think that the case Wolfgang mentioned can in theory indeed be a problem (I vaguely remember we had that problem when we implemented the first version of the infrastructure). However, I looked through the changes you made and I do not think the current version of the algorithm can ever run into this problem. Problems can only appear with algorithms of this sort:
```
for (particle: particles)
  for (cell: cells)
    if (particle in cell within tolerance)
      insert_particle_in_this_cell
```
However all algorithms you changed are like this:
```
for (particle: particles)
  for (cell: cells)
    if (particle in cell within tolerance)
    {
      cell = this cell
    }
    insert_particle_into_cell (cell)
```
So the particle is never duplicated, it will just be inserted into the **last** found cell if multiple cells are found (the same applies for the algorithm that checks if a particle is still in its old cell, we will just find a few more particles in their old cells, instead of looking for new cells for them, but never duplicate). Do you agree?

**Comment by mschreter** [comment]
I agree. Thanks for having a look into this!

**Comment by bangerth** [comment]
Thanks everyone for checking!

**Comment by blaisb** [comment]
Ohh this broke a test for us, so clearly this could have been an issue in the past :).
Thanks @mschreter  :)!

**Comment by bangerth** [comment]
On 2/25/24 12:34, Bruno Blais wrote:
> Ohh this broke a test for us, so clearly this could have been an issue in the 
> past :).

This means that what you have might also make for a nice test for deal.II if 
it happened to be easy enough to make independent of Lethe :-)

**Comment by blaisb** [comment]
> On 2/25/24 12:34, Bruno Blais wrote: Ohh this broke a test for us, so clearly this could have been an issue in the past :).
> This means that what you have might also make for a nice test for deal.II if it happened to be easy enough to make independent of Lethe :-)

I can try to reproduce in such a way. It's related to the advection of particles in a cylinder with manifolds, so clearly I think I can try to make a new version of it.
We have something like 100+ tests that include particles (some with up to 100k particles), so whenever there is a slight change to the particle library something pops up in Lethe.

**Pull request merged**